### PR TITLE
many markdown fixes

### PR DESCRIPTION
**Pull request by apatrida**
Don't escape within code blocks because HTML entities are not processed by markdown
Fence a triple backtick code block with 4 invisible spaces per line in case the code block contains backticks that might confuse the fencing
Check for backticks in inline code snippets and fence with more backticks than those present (which is the standard way to escape)
Change LF to <br/> or empty depending on the container (i.e. list items cannot have LF but rather can have <p> or <br/>, and tables as well in markdown cannot have LF but can have the <p> and <br/>
Fix bug where list items copied from includes add a line after the * and before the text (markdown parser generates a paragraph which is NOT intended to be a line break in the middle of a list item
Fix tables, they require a header row and the divider |---|---| style thing

should fix:  #71, #72 and others not yet reported

### CODE REVIEW

**Comment by apatrida** [comment]
This is somewhat hacky, because the generator API needs to be redone to do this better.

I might take a look at that in the future.  I think it might be best to generate more of a MarkDown ContentNode tree for everything, then use the normal rendering of that as HTML for that output, and for markdown render it from the tree.  Currently it is a mix of tree visiting and ad hoc writes to the output.

**Comment by apatrida** [comment]
I still need to update the unit tests, but if you have comments before then let me know.  All markdown unit tests fail.

**Comment by apatrida** [comment]
Ok, looks like the last remaining main issue (for those I have checked) is to change the signatures not to be in backtick blocks, but `<code>` blocks instead so that the links will work.

**Comment by apatrida** [comment]
a build of this is in my bintray for core and gradle plugin:
https://bintray.com/jaysonminard/kohesive/dokka/view

**Comment by apatrida** [comment]
test docs generated using this version:

https://github.com/kohesive/klutter/blob/master/docs/markdown/klutter-vertx3-jdk8/uy.klutter.vertx/index.md

https://github.com/kohesive/klutter/blob/master/docs/markdown/klutter-core-jdk6/index.md

**Comment by yole** [comment]
This is unfortunately very hard to review in this form. Would it be possible to split this into multiple pull requests which change one thing at a time? I'll leave some comments for this PR according to what I can see in this version of the code.

File: core/src/main/kotlin/Formats/HtmlFormatService.kt [hunk 1]
```diff
@@ -45,22 +49,20 @@ open class HtmlFormatService @Inject constructor(@Named("folders") locationServi
...
     }
 
-    override fun appendTable(to: StringBuilder, body: () -> Unit) {
+    override fun appendTable(to: StringBuilder, columnCount: Int, body: () -> Unit) {
         to.appendln("<table>")
         body()
         to.appendln("</table>")
     }
 
-    override fun appendTableHeader(to: StringBuilder, body: () -> Unit) {
```
Comment (yole) [inline core/src/main/kotlin/Formats/HtmlFormatService.kt]: Why did you remove this method?

File: core/src/main/kotlin/Formats/StructuredFormatService.kt [hunk 2]
```diff
@@ -55,6 +56,29 @@ abstract class StructuredFormatService(locationService: LocationService,
         return StringBuilder().apply { formatText(location, content, this, listKind) }.toString()
     }
 
+    fun ContentNode.ignoreParagraphAtTop(): ContentNode {
```
Comment (yole) [inline core/src/main/kotlin/Formats/StructuredFormatService.kt]: This logic is already handled by the buildInlineContentTo() function; doesn't need to be duplicated.

File: core/src/main/kotlin/Formats/MarkdownFormatService.kt [hunk 3]
```diff
@@ -44,44 +118,57 @@ open class MarkdownFormatService(locationService: LocationService,
...
+    override fun appendBlockCode(to: StringBuilder, language: String, bodyAsLines: ()->List<String>) {
+        to.append(allowedLf().requiredWhitespace)
+
+        val codeTags = allowedLf().inlineCode
+        if (codeTags.start.startsWith('`')) {
+            changeLf(LfPriority.Code) {
+                // this does not adjust line feeds, a code block in a list or table will just break formatting
+                to.appendln(if (language.isEmpty()) "```" else "``` $language")
+                // use 4 zero width spaces which act as a code fence escaping any use of ``` within the code block
+                to.appendln(bodyAsLines().map { "​\u200B\u200B\u200B\u200B$it" }.joinToString("\n"))
```
Comment (yole) [inline core/src/main/kotlin/Formats/MarkdownFormatService.kt]: Why exactly is this needed? And are you sure this is going to be processed as you expect by all Markdown renderers?

File: core/src/main/kotlin/Formats/HtmlFormatService.kt (hunk 1 above)
Comment (apatrida) [inline core/src/main/kotlin/Formats/HtmlFormatService.kt]: It wasn't used and to add it causes a problem for Markdown I didn't want to solve unless this is every brought back to be used.  Then ONLY if it is really needed, do the hard work to solve the problem.

File: core/src/main/kotlin/Formats/MarkdownFormatService.kt (hunk 3 above)
Comment (apatrida) [inline core/src/main/kotlin/Formats/MarkdownFormatService.kt]: Few markdown renderers support headerless tables, and common mark the parser you are using doesn't either.  It is currently being discussed how headerless tables would be in markdown and there is no consensus.  The table format that was there before was the rare unsupported case of having no headers and no `---|---|---` row.

File: core/src/main/kotlin/Formats/StructuredFormatService.kt (hunk 2 above)
Comment (apatrida) [inline core/src/main/kotlin/Formats/StructuredFormatService.kt]: I'll look at that, previously everything was getting extra paragraphs that shouldn't causing line breaks which broke the markdown and made it invalid.

File: core/src/main/kotlin/Formats/MarkdownFormatService.kt (hunk 3 above)
Comment (yole) [inline core/src/main/kotlin/Formats/MarkdownFormatService.kt]: I mean, why are these four zero-width spaces needed?

**Comment by apatrida** [comment]
In general most of the changes work together, the rendering model is really hard for markdown.  It was built for HTML then tried to adapt to Markdown.  Would be better to build a complete tree using the Markdown AST no matter what renderer you are using, then render from there.  In the HTML case the default markdown-to-html rendering could be used instead of a custom renderer.

File: core/src/main/kotlin/Formats/MarkdownFormatService.kt (hunk 3 above)
Comment (apatrida) [inline core/src/main/kotlin/Formats/MarkdownFormatService.kt]: Ah code fencing.  It is because if you use backticks inside of code it can break the code block.  

You can create a code block alternatively using 4 white-space characters.  

So by using 4 zero space white space characters it ensures no matter what fencing is used, that backticks within the code block are ignored.

It is basically escaping everything each code block line.  

It is a work around until the "add more back ticks per fence" method is supported by more markdown renderers.

Comment (apatrida) [inline core/src/main/kotlin/Formats/MarkdownFormatService.kt]: (assuming you don't have a parser that doesn't treat all whitespace the right way, then this would do nothing for that parser)

**Comment by yole** [comment]
Rewriting Dokka to use the Markdown AST for everything isn't going to happen, but it's definitely possible to adapt the content block structure so that it would work better with Markdown, so if you can point to specific places which cause problems, please do.

**Comment by apatrida** [comment]
Maybe the rendering methods could change then instead to be workable for both.  I had to force-fit the HTML rendering style pattern to work with Markdown.

Another option is to use html tags in the markdown, but it isn't clear which renderers support which basic tags, and if a person might have them turned off.  

So these changes try to use the standard markdown language unless it would not work at all in some cases then use html tags which might be stripped (which is fine, it does best attempt).  

The current model also wanted to escape things before building strings and was escaping for HTML and not for markdown, so that was a problem too.  

Basically this commit represents the basic "make everything work right" at least from the renderers I checked including Github's.  (likely the most common use of this is Github yes?)

**Comment by yole** [comment]
I've been looking at this some more, and I think that it would be much more valuable to have testdata on which the current formatter is producing incorrect results, rather than the changes to the formatter itself. I think that the LfPriority mechanism you've implemented is a little too complicated, and can be easier handled by separating MarkdownTextBuilder (which knows the current state of the output Markdown) from MarkdownFormatService. However, in the current state, where you broke all tests and didn't provide updated testdata, I have no way to ensure that such a refactoring would solve the problems that motivated you to make these changes.

**Comment by yole** [comment]
Also it looks like most Markdown renderers support headerless tables, and only GFM requires a header row (and renders the current headers without text in a fairly ugly way). I think it would be cleaner to provide the text for table headers to the rendering service and to provide a separate GFM output format, which would render the headers as supported by GFM.

**Comment by yole** [comment]
I decided to approach the problem differently and did a fairly massive rewrite that produces a much cleaner output (fewer redundant line breaks, for one) with a simpler implementation. This should fix all the issues that you're addressing with this pull request, aside from multiple backticks inside a code block (you're welcome to file a separate issue about this). I'd appreciate it if you could test the master version on your library and let me know if there are any issues with formatting. Please use "-format gfm" if you plan to publish the content on GitHub Pages.

In any case, thanks for your contribution!
